pubtools/_quay/security_manifest_pusher.py

```diff
@@ -429,13 +432,75 @@ def merge_and_push_security_manifest(
                 )
                 full_security_manifest_path = image_manifest.security_manifest_path
 
+            repos_attestation_mapping[repo] = full_security_manifest_path
             self.cosign_attest_security_manifest(
                 full_security_manifest_path,
                 image_ref,
                 self.target_settings.get("cosign_rekor_url", None),
                 self.target_settings.get("cosign_sbom_skip_upload_rekor", False),
             )
 
+        return repos_attestation_mapping
+
+    def push_manifest_list_security_manifests(
+        self,
+        source_ref: str,
+        full_repos_attestations_mapping: dict[str, list[str]],
+        dir_path: str,
+    ) -> None:
+        """
+        Push all arch attestations to the manifest list digest.
+
+        This ensures that getting an attestation of an images specified via tag returns a result.
+        The result will be all arch attestations in a JSONL format.
+
+        Args:
+            source_ref (str):
+                Reference to the multiarch image.
+            full_repos_attestations_mapping (dict[str, list[str]]):
+                Mapping of repos and all arch attestations they contain.
+            dir_path (str):
+                Directory path where artifacts will be created.
+        """
+        ml_digest = self.src_quay_client.get_manifest_digest(
+            source_ref, self.src_quay_client.MANIFEST_LIST_TYPE
+        )
+
+        for repo in full_repos_attestations_mapping:
+            image_ref = f"{repo}@{ml_digest}"
+            LOG.info(
+                f"Adding attestations to multiatch image {image_ref}. The number of "
+                f"attestations is {len(full_repos_attestations_mapping[repo])}"
+            )
+
+            attestation_file = os.path.join(dir_path, f"attestation_{uuid.uuid4().hex}.json")
+            attestation_exist = self.cosign_get_existing_attestation(
+                image_ref,
+                attestation_file,
+                self.target_settings.get("cosign_rekor_url", None),
+                self.target_settings.get("cosign_sbom_skip_verify_rekor", False),
+            )
+            # Trying to determine if a multiarch attestation has changed by the new push is
+            # too complicated. Let's always remove it and replace it by a new one, even if
+            # there are no meaningful changes.
+            if attestation_exist:
+                LOG.info(
+                    f"Multiarch image {image_ref} already has an attestation. "
+                    "It will be replaced by a new attestation containing updated information."
+                )
+                self.delete_existing_attestation(image_ref, dir_path)
```

Comment:
I wonder if it's ok to remove existing attestations. As we also don't remove existing manifests from manifest list we overwrite because they can be included in other manifest list.

Reply:
The old multiarch attestation is removed and replaced with a new one containing all up-to-date arch attestations. Is there a problem with this approach that I have missed?

Reply:
My point is old attestation can be valid if a manifest in old ML is referenced by other tag

Reply:
That's a good point, I haven't considered ML merging.  I've changed the logic to create the ML attestations based on their arch digests. I think that should guarantee that each pushed multiarch tag should have the correct attestations.

Reply:
I still think it won't resolve the problem. Let say you have manifest list L with manifests l1,l2,l3 and manifest list K with manifests k1, k2, l3. If you push new manifest list L2 it removes all attestation for l1,l2,l3 isn't that correct? Which means manifest list K will have manifest l3 without attestation. If I understand correctly, if you push L first and then K, l3 already has attestation and even when you recreate it at the time of push of K it will still be shared for l3 in K and L - as it's referenced by digest not by tag+digest

Reply:
The attestation images are not removed in such workflows, they can only be removed when explicitly deleting tags, with image_untagger.py. Image untagger makes checks so that an attestation is only removed if all the images of a given digest are removed in a repository. 

I suppose it's possible that a digest can be overwritten in a repo and an SBOM image would be left "dangling", no longer being associated with any tagged image. I don't know if it's possible in practice though, I think there are always non-floating tags being pushed which will never be overwritten.

Since the attestation images are never removed (unless explicitly deleted), I think it's safe to construct the ML attestation by gathering all its arch attestations via arch digests.

What do you think?

Reply:
yeah, in that case it makes sense